### PR TITLE
Sema: Rewrite partial applications into closures

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes rdar://21289579 / https://bugs.swift.org/browse/SR-75.

There's a big follow-up cleanup, which is removing curry thunks and assorted machinery from SILGen. I'll land that separately.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

File: test/Index/roles.swift [hunk 1]
```diff
@@ -175,9 +175,9 @@ class AClass {
 }
 
 let _ = AClass.foo
-// CHECK: [[@LINE-1]]:16 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref | rel: 0
+// CHECK: [[@LINE-1]]:16 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref,Call | rel: 0
 let _ = AClass(x: 1).foo
-// CHECK: [[@LINE-1]]:22 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref | rel: 0
+// CHECK: [[@LINE-1]]:22 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref,Call | rel: 0
```
Comment (akyrtzi) [inline test/Index/roles.swift]: Marking these as "call sites" is inaccurate, could you correct this?

File: test/Index/roles.swift [hunk 2]
```diff
@@ -352,7 +352,7 @@ let otherInstance = AStruct(x: 1)
 
 _ = AStruct.init(x:)
 // CHECK: [[@LINE-1]]:18 | instance-property/Swift | x | [[AStruct_x_USR]] | Ref | rel: 0
-// CHECK: [[@LINE-2]]:5 | struct/Swift | AStruct | [[AStruct_USR]] | Ref | rel: 0
```
Comment (akyrtzi) [inline test/Index/roles.swift]: Is the reference to `AStruct` in this line not reported anymore?

File: lib/Sema/CSApply.cpp [hunk 3]
```diff
@@ -604,7 +606,10 @@ namespace {
       if (isa<AbstractFunctionDecl>(member)) {
         // For functions, close the existential once the function
         // has been fully applied.
+        assert(member->getDeclContext()->isTypeContext());
         return 2;
+      } else if (auto *elt = dyn_cast<EnumElementDecl>(member)) {
+        return elt->getParameterList() ? 2 : 1;
```
Comment (CodaFi) [inline lib/Sema/CSApply.cpp]: Can the brains of this function just be `ValueDecl::getNumCurryLevels()`?  The assertions are still good to keep.

File: lib/Sema/ConstraintSystem.cpp [hunk 4]
```diff
@@ -1509,16 +1488,35 @@ ConstraintSystem::getTypeOfMemberReference(
   }
 
   // Compute the type of the reference.
-  Type type;
+  Type type = openedType;
+
+  if (!outerDC->getSelfProtocolDecl()) {
```
Comment (CodaFi) [inline lib/Sema/ConstraintSystem.cpp]: This is like the third or fourth time this block appears.  Can we centralize this?  One of them is going to fall out of sync one day...

**Review by CodaFi — APPROVED** [review]
I've only got nits.  The approach looks fantastic.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by tkremenek** [comment]
@swift-ci clean test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
I dusted this off and fixed all the merge conflicts. Still need to address Robert's review feedback and a couple of other nits.

File: lib/Sema/CSApply.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/Sema/CSApply.cpp]: Yup! Thanks.

File: lib/Sema/ConstraintSystem.cpp (hunk 4 above)
Comment (slavapestov) [reply lib/Sema/ConstraintSystem.cpp]: I only found two, but I'll extract it into a new method on ValueDecl.

**Comment by akyrtzi** [comment]
Just a reminder that the output changes in `test/Index/roles.swift` are technically regressions and it would be great to address them before merging 🙏

**Comment by slavapestov** [comment]
@akyrtzi Yep, they're already addressed in the latest version!

File: test/Index/roles.swift (hunk 1 above)
Comment (slavapestov) [reply test/Index/roles.swift]: This test passes with no changes in the latest version of the commit.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

File: CHANGELOG.md [hunk 5]
```diff
@@ -46,6 +46,19 @@ Swift Next
     // OK, <U where U: Equatable> has broader visibility than <T where T == Int>
     override func foo() where U: Equatable { ... } 
   }
+
+* [SR-75][]:
+
+  Unapplied references to protocol methods methods are now supported. Previously this
```
Comment (benrimmington) [inline CHANGELOG.md]: `methods methods` ➡️ `methods`

```diff
-Unapplied references to protocol methods methods are now supported. Previously this
+Unapplied references to protocol methods are now supported. Previously this
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -46,6 +46,19 @@ Swift Next
...
+  only worked for methods defined in structs, enums and classes.
+
+  ```swift
+  protocol Cat {
+    func play(catToy: Toy)
+  }
+
+  let fn = Cat.play
```
Comment (benrimmington) [inline CHANGELOG.md]: [Require parameter names when referencing to functions](https://forums.swift.org/t/require-parameter-names-when-referencing-to-functions/27048)?

```diff
-let fn = Cat.play
+let fn = Cat.play(catToy:)
```

File: CHANGELOG.md (hunk 5 above)
Comment (slavapestov) [reply CHANGELOG.md]: Thanks!

File: CHANGELOG.md (hunk 6 above)
Comment (slavapestov) [reply CHANGELOG.md]: Not actually required, but I agree it makes the example more clear.

**Comment by davezarzycki** [comment]
Hi @slavapestov – Congratulations on landing this and fixing a long standing issue. That's awesome! Do you know of (or foresee) any case where a `DotSyntaxBaseIgnoredExpr` would have an implicit autoclosure on the RHS that is NOT the thunk logic created by this pull request? Thanks in advance – Dave

**Comment by slavapestov** [comment]
@davezarzycki That's a good question. It may very well be that the answer is "no". Do you have a further simplification in mind?

**Comment by davezarzycki** [comment]
I don’t have a simplification in mind although it does seem like it could be simplified.

FWIW – This change broke a research branch I’m working on and I wanted to know how paranoid my misc diagnostic should be after this pull request. I think I know now. Thanks

**Pull request merged**